spark/src/main/scala/io/delta/tables/DeltaTable.scala

```diff
@@ -562,6 +562,69 @@ class DeltaTable private[tables](
           TableFeatureProtocolUtils.FEATURE_PROP_SUPPORTED))
     toDataset(sparkSession, alterTableCmd)
   }
+
+  /**
+   * Create a shallow clone of this DeltaTable to a new location.
+   *
+   * @param target The path or table name to create the clone
+   * @param create Whether to create the clone if no table exists
+   * @param replace Whether to replace anything at the destination with the clone
+   * @param ifNotExists Whether to ignore the operation if a table already exists at the path
+   * @param tableProperties Any table properties to override in the clone
+   * @since 4.0.0
+   */
+  def clone(
+      target: String,
+      create: Boolean = true,
+      replace: Boolean = false,
+      ifNotExists: Boolean = false,
+      tableProperties: Map[String, String] = Map.empty): DataFrame = {
+    executeClone(table, target, create, replace, ifNotExists, tableProperties)
```

Comment:
Hey @Kimahriman, I think the APIs you adding are good, I have been brainstorming about the clone APIs, and I think that:

1. Users should not be able to shoot themselves in the foot, and should be succeed with their API calls, i.e. APIs should be robust.
- https://github.com/delta-io/delta/blob/992be8412c1e0b843ef9881995d63800f95f9f02/spark/src/main/scala/io/delta/sql/parser/DeltaSqlParser.scala#L288 In this case, if someone specify `create = false` and `ifNotExists = true`, then they would get this error `DELTA_PARSING_MUTUALLY_EXCLUSIVE_CLAUSES`, so these 2 arguments `create` and `ifNotExists` should not exist at the same time.
- We cannot have `create = false` and `replace = false`, so these 2 arguments also should not exist at the same time.
- We cannot have `replace = false` and `ifNotExists = .`, this combination doesn't exist.

That means we should only keep one argument, I think we should keep `replace` as a variable while `ifNotExists = false` while `isCreate = true`, because
- `create` and `replace` as default while `ifNotExists` as a variable would limit the functionality drastically.
- I think we would always want the target table to exist (so `create = true`) than always replacing it (`replace = true`) or not replacing it (`replace = false`).

2. We want the Scala APIs to also be Java compatible, and similar to other Delta Table's Scala APIs, we should create the Scala APIs with no default, and do something like this.

`f(a, b, c): executeF(a, b, c)`
`f(a, b): executeF(a, b, defaultC)`
`f(a): executeF(a, defaultB, defaultC)`

https://docs.delta.io/latest/api/java/spark/index.html
https://docs.delta.io/latest/api/scala/spark/io/delta/tables/DeltaTable.html

Java doesn't have default values for arguments, like Scala here, and it would create `f$default$x` like you saw - https://github.com/delta-io/delta/pull/3392#discussion_r1717469314, and so it would not be Java compatible.

3. `tableProperties` should be called `properties` for brevity

What do you think? I'm happy to make the changes for you for these 3 clone functions if you agree.

Reply:
I was mostly trying to copy what little I could find on the current Databricks implementation: https://docs.databricks.com/en/delta/clone.html#example-clone-syntax. 

We could keep multiple arguments if we wanted, just make them all required. But you're basically thinking something like

```scala
def clone(target: String): DataFrame = {
  clone(target, false)
}

// This could have multiple required flags if desired
def clone(target: String, replace: Boolean): DataFrame = {
  clone(target, replace, Map.empty)
}

def clone(target: String, replace: Boolean, properties: Map[String, String]): DataFrame = {
  executeClone(...)
}
```
?

Reply:
@Kimahriman 

Yes, [convertToDelta](https://github.com/delta-io/delta/blob/597950c5ff5cf0aa2b76a3d5fba9da58fdba332e/spark/src/main/scala/io/delta/tables/DeltaTable.scala#L624) and [vacuum](https://github.com/delta-io/delta/blob/597950c5ff5cf0aa2b76a3d5fba9da58fdba332e/spark/src/main/scala/io/delta/tables/DeltaTable.scala#L100) are good examples of this.

"// This could have multiple required flags if desired" What do you mean by this sorry? What are you referring to by "flags"?

What I'm proposing are:

`def clone(target: String, replace: Boolean, properties: Map[String, String]): DeltaTable`
`def clone(target: String, replace: Boolean): DeltaTable`
`def clone(target: String): DeltaTable`
`def cloneAtVersion(version: Int, target: String, replace: Boolean, properties: Map[String, String]): DeltaTable`
`def cloneAtVersion(version: Int, target: String, replace: Boolean): DeltaTable`
`def cloneAtVersion(version: Int, target: String): DeltaTable`
`def cloneAtTimestamp(timestamp: String, target: String, replace: Boolean, properties: Map[String, String]): DeltaTable`
`def cloneAtTimestamp(timestamp: String, target: String, replace: Boolean): DeltaTable`
`def cloneAtTimestamp(timestamp: String, target: String): DeltaTable`

Reply:
I just meant if we wanted to keep create or ifNotExists as well

Reply:
Updated